src/react/elements.js

```diff
@@ -0,0 +1,147 @@
+import { Get } from "../methods/index.js";
+import { getReactSymbol, objectHasNoPartialKeyAndRef, deleteRefAndKeyFromProps } from "./utils.js";
+import * as t from "babel-types";
+import { computeBinary } from "../evaluators/BinaryExpression.js";
+import { CompilerDiagnostic, FatalError } from "../errors.js";
+
+function createPropsObject(
+  realm: Realm,
+  type: Value,
+  config: ObjectValue | AbstractValue | AbstractObjectValue | NullValue,
+  children: Value
+) {
+  let defaultProps = type instanceof ObjectValue ? Get(realm, type, "defaultProps") : realm.intrinsics.undefined;
+  let props = Create.ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
+  // start by having key and ref deleted, if they actually exist, they will be add later
+  deleteRefAndKeyFromProps(realm, props);
+  let key = realm.intrinsics.null;
+  let ref = realm.intrinsics.null;
+
+  const setProp = (name: string, value: Value): void => {
+    if (name === "children") {
+      invariant(props instanceof ObjectValue);
+      Properties.Set(realm, props, "children", value, true);
+    } else if (name === "key" && value !== realm.intrinsics.null) {
+      key = computeBinary(realm, "+", realm.intrinsics.emptyString, value);
+    } else if (name === "ref") {
+      ref = value;
+    } else if (name !== "__self" && name !== "__source") {
+      invariant(props instanceof ObjectValue);
+      Properties.Set(realm, props, name, value, true);
+    }
+  };
+
+  if (
+    (config instanceof AbstractObjectValue && config.isPartialObject()) ||
+    config instanceof AbstractValue ||
+    (config instanceof ObjectValue && config.isPartialObject() && config.isSimpleObject())
+  ) {
+    // if we have defaultProps, we need to create a new merge of the objects along with our config
+    if (defaultProps !== realm.intrinsics.undefined || children !== realm.intrinsics.undefined) {
+      if (objectHasNoPartialKeyAndRef(realm, config)) {
+        let args = [];
+        if (defaultProps !== realm.intrinsics.undefined) {
+          args.push(defaultProps);
+        }
+        args.push(config);
+        // create a new props object that will be the target of the Object.assign
+        props = Create.ObjectCreate(realm, realm.intrinsics.ObjectPrototype);
+        // as this is "props" that is abstract, we need to make it partial and simple
+        props.makePartial();
+        props.makeSimple();
+        // props objects also don't have a key and ref, so we remove them
+        deleteRefAndKeyFromProps(realm, props);
+        let types = new TypesDomain(FunctionValue);
+        let values = new ValuesDomain();
+
+        // get the global Object.assign
+        let globalObj = Get(realm, realm.$GlobalObject, "Object");
+        invariant(globalObj instanceof ObjectValue);
+        let objAssign = Get(realm, globalObj, "assign");
+        invariant(realm.generator);
+
+        realm.generator.derive(types, values, [objAssign, props, ...args], ([methodNode, ..._args]) => {
```

Comment:
Should this use `AbstractValue.createTemporalFromBuildFunction`?
I don't understand how to choose between that and `derive()`.

Reply:
Only AbstractValue should use derive. Any other code should use factories in AbstractValue to create abstract values.